File: leetcode-js/264-ugly-number-ii.js
/**
 * @param {number} n
 * @return {number}
 */
var nthUglyNumber = function(n) {
    'use strict';

    const queue = [1];
    let i2 = 0;
    let i3 = 0;
    let i5 = 0;

    while (queue.length < n) {
        const m2 = queue[i2] * 2;
        const m3 = queue[i3] * 3;
        const m5 = queue[i5] * 5;
        const m = Math.min(m2, m3, m5);

        if (m === m2) {
            i2 += 1;
        }

        if (m === m3) {
            i3 += 1;
        }

        if (m === m5) {
            i5 += 1;
        }

        queue.push(m);
    }

    return queue[queue.length - 1];
};

console.log(nthUglyNumber(7));

File: leetcode-java/264-ugly-number-ii.java
public class Solution {
    public int nthUglyNumber(int n) {
        int[] ugly = new int[n];

        int index2 = 0;
        int index3 = 0;
        int index5 = 0;

        int factor2 = 1;
        int factor3 = 1;
        int factor5 = 1;

        int next = 1;

        for (int i = 0; i < n; i++) {
            ugly[i] = next;

            if (ugly[i] == factor2) {
                factor2 = ugly[index2] * 2;
                index2++;
            }

            if (ugly[i] == factor3) {
                factor3 = ugly[index3] * 3;
                index3++;
            }

            if (ugly[i] == factor5) {
                factor5 = ugly[index5] * 5;
                index5++;
            }

            next = Math.min(factor2, Math.min(factor3, factor5));
        }

        return ugly[n - 1];
    }
}
